### PR TITLE
Implement Capacity Metric and W&B Histogram Logging

### PR DESCRIPTION
**Pull request by HoagyC**
Added calc_capacities to compute feature capacities as per [arXiv:2210.01892](https://arxiv.org/pdf/2210.01892.pdf) and wandb_capacities_histogram for logging the results in Weights & Biases. Also added unit tests. Trying to get a feel for metrics and the codebase before looking at adding metrics abstraction.

### CODE REVIEW

File: sparse_autoencoder/train/metrics/capacity.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        >>> orthogonal_caps = calc_capacities(orthogonal_features)
+        >>> assert torch.allclose(orthogonal_caps, torch.tensor([1., 1., 1.]))
+
+    Args:
+        features: A 2D tensor of features, where each row is a feature.
+
+    Returns:
+        A 1D tensor of capacities, where each element is the capacity of the corresponding feature.
+    """
+    squared_dot_products = (features @ features.T).pow(2)
```
Comment (alan-cooney) [inline sparse_autoencoder/train/metrics/capacity.py:34]: Nit: worth doing in einops to make more readable

File: sparse_autoencoder/train/metrics/tests/test_capacities.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    assert torch.allclose(
+        orthogonal_capacities, torch.tensor([1.0, 1.0, 1.0])
+    ), "Orthogonal features should have capacity 1.0."
+    assert torch.allclose(
+        identical_capacities, torch.ones(3) / 3
+    ), "Identical features should have capacity 1/3."
+    assert torch.allclose(
+        intermediate_capacities, torch.tensor([2 / 3, 2 / 3, 1.0]), rtol=1e-3
+    ), "Capacity calculation is incorrect."
```
Comment (alan-cooney) [inline sparse_autoencoder/train/metrics/tests/test_capacities.py:29]: Tests look good, but it's best practice to test one thing per test - this way it's easy to debug when tests fail as you know exactly what has gone wrong. With pytest the tool to do this is `@pytest.mark.parametrize` https://docs.pytest.org/en/7.3.x/how-to/parametrize.html

File: sparse_autoencoder/train/metrics/tests/test_capacities.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        3,
+    ], "Histogram is incorrect."
```
Comment (alan-cooney) [inline sparse_autoencoder/train/metrics/tests/test_capacities.py:58]: I think you're sort of doing a manual snapshot here.

I'd suggest using snapshottest as it's included in the lib - https://pypi.org/project/snapshottest/ - although note that currently ruff will complain unless you add docstrings after you've generated the snapshot files

**Review by alan-cooney — CHANGES_REQUESTED** [review]
Thanks for adding this!

Mostly looks good - a few suggestions in the comments

**Review by alan-cooney — APPROVED** [review]
Nice work on this - looks good!

I made a few minor changes directly, mostly just running `poe test --snapshot-update` and also removing the legacy snapshottest stuff
